File: backend-ci/gestion/services/lieu_service.py
from django.db import transaction
from django.shortcuts import get_object_or_404
from gestion.models.lieu import Lieu


class LieuService:

    @staticmethod
    @transaction.atomic
    def create(data):
        """Créer un nouveau lieu"""
        lieu = Lieu.objects.create(
            nom=data['nom'],
            categorie=data.get('categorie', 'AUTRE'),
            latitude=data['latitude'],
            longitude=data['longitude'],
            description=data.get('description', ''),
            adresse=data.get('adresse', '')
        )
        return lieu

    @staticmethod
    def get_all():
        """Récupérer tous les lieux actifs"""
        return Lieu.objects.filter(actif=True)

    @staticmethod
    def get_by_id(lieu_id):
        """Récupérer un lieu par son ID"""
        return get_object_or_404(Lieu, id=lieu_id)

    @staticmethod
    def get_by_categorie(categorie):
        """Récupérer les lieux par catégorie"""
        return Lieu.objects.filter(categorie=categorie, actif=True)

    @staticmethod
    def update(lieu, data):
        """Mettre à jour un lieu"""
        for key, value in data.items():
            setattr(lieu, key, value)
        lieu.save()
        return lieu

    @staticmethod
    def delete(lieu):
        """Supprimer un lieu (soft delete)"""
        lieu.actif = False
        lieu.save()

File: frontend-ci/src/services/lieu.service.ts
/**
 * Service pour les lieux
 * Correspond à gestion/services/lieu_service.py
 */

import apiClient from './api';
import { Lieu, SaveLieuDTO, UpdateLieuDTO, CreateResponse } from '../types';
import { CategorieLieu } from '../types/enums';

const ENDPOINTS = {
  BASE: '/lieux/',
  DETAIL: (id: number) => `/lieux/${id}/`,
  BY_CATEGORIE: (categorie: CategorieLieu) => `/lieux/?categorie=${categorie}`,
};

export const lieuService = {
  /**
   * Récupérer tous les lieux actifs
   */
  async getAll(): Promise<Lieu[]> {
    const response = await apiClient.get<Lieu[]>(ENDPOINTS.BASE);
    return response.data;
  },

  /**
   * Récupérer un lieu par son ID
   */
  async getById(id: number): Promise<Lieu> {
    const response = await apiClient.get<Lieu>(ENDPOINTS.DETAIL(id));
    return response.data;
  },

  /**
   * Récupérer les lieux par catégorie
   */
  async getByCategorie(categorie: CategorieLieu): Promise<Lieu[]> {
    const response = await apiClient.get<Lieu[]>(ENDPOINTS.BY_CATEGORIE(categorie));
    return response.data;
  },

  /**
   * Créer un nouveau lieu
   */
  async create(data: SaveLieuDTO): Promise<CreateResponse> {
    const response = await apiClient.post<CreateResponse>(ENDPOINTS.BASE, data);
    return response.data;
  },

  /**
   * Mettre à jour un lieu
   */
  async update(id: number, data: UpdateLieuDTO): Promise<Lieu> {
    const response = await apiClient.put<Lieu>(ENDPOINTS.DETAIL(id), data);
    return response.data;
  },

  /**
   * Supprimer un lieu (soft delete)
   */
  async delete(id: number): Promise<void> {
    await apiClient.delete(ENDPOINTS.DETAIL(id));
  },
};

export default lieuService;
